src/palindrom-network-channel.js

```diff
@@ -0,0 +1,362 @@
+        onSend && (this.onSend = onSend);
+        onConnectionError && (this.onConnectionError = onConnectionError);
+        onFatalError && (this.onFatalError = onFatalError);
+        onStateChange && (this.onStateChange = onStateChange);
+        onSocketOpened && (this.onSocketOpened = onSocketOpened);
+
+        Object.defineProperty(this, 'useWebSocket', {
+            get: function() {
+                return useWebSocket;
+            },
+            set: newValue => {
+                useWebSocket = newValue;
+
+                if (newValue == false) {
+                    if (this._ws) {
+                        this._ws.onclose = function() {
+                            //overwrites the previous onclose
+                            this._ws = null;
+                        };
+                        this._ws.close();
+                    }
+                    // define wsUrl if needed
+                } else if (!this.wsUrl) {
+                    this.wsUrl = toWebSocketURL(this.remoteUrl.href);
+                }
+                return useWebSocket;
+            }
+        });
+    }
+    get useWebSocket() {
+        return this._useWebSocket;
+    }
+    set useWebSocket(newValue) {
+        this._useWebSocket = newValue;
+
+        if (newValue == false) {
+            if (this._ws) {
+                this._ws.onclose = () => {
+                    //overwrites the previous onclose
+                    this._ws = null;
+                };
+                this._ws.close();
+            }
+            // define wsUrl if needed
+        } else if (!this.wsUrl) {
+            this.wsUrl = toWebSocketURL(this.remoteUrl.href);
+        }
+        return this.useWebSocket;
+    }
+
+    async _establish(reconnectionPendingData = null) {
+        const data = await this._fetch(
+            reconnectionPendingData ? 'PATCH' : 'GET',
+            this.remoteUrl.href + (reconnectionPendingData ? '/reconnect' : ''),
+            'application/json',
+            reconnectionPendingData
+        );
+        if (this.useWebSocket) {
+            this.webSocketUpgrade(this.onSocketOpened);
+        }
+        return data;
+    }
+
+    /**
+     * Send any text message by currently established channel
+     * @TODO: handle readyState 2-CLOSING & 3-CLOSED (tomalec)
+     * @param  {String} msg message to be sent
+     * @return {PalindromNetworkChannel}     self
+     */
+    async send(msg) {
+        // send message only if there is a working ws connection
+        if (this.useWebSocket && this._ws && this._ws.readyState === 1) {
+            this._ws.send(msg);
+            this.onSend(msg, this._ws.url,'WS');
+        } else {
+            const url = this.remoteUrl.href;
+            const method = 'PATCH';
+            const data = await this._fetch(
+                method,
+                url,
+                'application/json-patch+json',
+                msg
+            );
+            this.onReceive(data, url, method);
```

Comment:
This changes the API as `data` now is a parsed JSON object, not a string.
@warpech are we fine with such API change?

Reply:
Not really. That's the mistake that is causing TeamCity tests to fail

Reply:
@warpech are we fine with changing this API as well, or should I bring back the string (should be pretty simple)?

Reply:
I could not find a single use of `onReceive` in Starcounter and Palindrom orgs, so I think it is fine to change it.

However, could you please go through the `docs/*` and update the API descriptions to include the new types.

We will use the updated docs as the basis for release notes.

Reply:
You should look for `onPatchReceived` and `handleRemoteChange` as those are public APIs

and I found a few:
https://github.com/Starcounter/PalindromBenchmark/blob/4af1db9474dd279931216f594552fa05980b33c7/src/PalindromBenchmark/wwwroot/Benchmark/assets/BenchmarkTests.js#L130 (does not use data)
https://github.com/Starcounter/AppStoreClient/blob/1a40772efb6be455f7db14a70d006444a29a4152/src/AppStoreClient/wwwroot/AppStoreClient/viewmodels/AppStoreAppsPage.html#L116 uses data, but not sure if we should bother about it
https://github.com/Starcounter/DevTools/blob/1e2a365312eb2d51ed5930277d4c4f92a7e80d67/src/palindrom-js-listener.js#L137 seems to use outdated API anyway

Reply:
https://github.com/Starcounter/HeadsOmni/blob/994a14d138783ac9059e0a4c2aa9bd3a331f8ee5/benchmark/core/lib/engine_palindrom.js#L198
https://github.com/Starcounter/appsground-performance-lab/blob/0f9e6acd9668cbdb89bcde7053c5b7bda1c9cd1c/sys/palindrom-client/palindrom-client.html#L415
https://github.com/Starcounter/PalindromBenchmark/blob/4af1db9474dd279931216f594552fa05980b33c7/src/PalindromBenchmark/wwwroot/Benchmark/assets/benchmarkLogic.js#L105

Reply:
Now I'm confused old tests seem to check `onPatchReceived` for an object https://github.com/Palindrom/Palindrom/pull/227/files#diff-2989c7b65809bbd4b69466ec6d71ebd2L60

Reply:
Ah ok, now I got it seems we had an inconsistent API previously
`onPatchReceived`'s first argument was an object for WS messages and a string for HTTP calls (establish, `getPatchUsingHTTP`, and `send` when useWebSocket is set to `false`)

I think we should keep it consistent and expose parsed object in all cases. It will reduce number of `JSON.parse`s, and allow getting reference to the actual object given in patch `value` (which could be good and bad).

Then if needed we would have to adopt the other code that uses Palindrom's `onPatchReceived` expecting it to be a string.

I'll push changes to docs.